src/rrules/avoiding_non_differentiable_code.jl

```diff
@@ -14,25 +14,190 @@ end
+end
+
+@is_primitive MinimalCtx Tuple{
+    typeof(Core._call_latest),
+    typeof(handle_message),
+    Any,
+    Base.CoreLogging.LogLevel,
+    String,
+    Module,
+    Symbol,
+    Symbol,
+    String,
+    Int64,
+}
+function rrule!!(
+    ::CoDual{typeof(Core._call_latest)},
+    ::CoDual{typeof(CoreLogging.handle_message)},
+    logger::CoDual,
+    loglevel::CoDual{Base.CoreLogging.LogLevel},
+    message::CoDual{String},
+    _module::CoDual{Module},
+    group::CoDual{Symbol},
+    id::CoDual{Symbol},
+    file::CoDual{String},
+    line::CoDual{Int64},
+)
+    result = Core._call_latest(
+        CoreLogging.handle_message,
+        logger.x,
+        loglevel.x,
+        message.x,
+        _module.x,
+        group.x,
+        id.x,
+        file.x,
+        line.x,
+    )
+    return zero_fcodual(result), NoPullback()
+end
+
+@is_primitive MinimalCtx Tuple{
+    typeof(Core._call_latest),
+    typeof(Core.kwcall),
+    <:NamedTuple,
+    typeof(CoreLogging.handle_message),
+    Any,
+    Base.CoreLogging.LogLevel,
+    String,
+    Module,
+    Symbol,
+    Symbol,
+    String,
+    Int64,
+}
+function rrule!!(
+    ::CoDual{typeof(Core._call_latest)},
+    ::CoDual{typeof(Core.kwcall)},
+    kwargs::CoDual{<:NamedTuple},
+    ::CoDual{typeof(CoreLogging.handle_message)},
+    logger::CoDual,
+    loglevel::CoDual{Base.CoreLogging.LogLevel},
+    message::CoDual{String},
+    _module::CoDual{Module},
+    group::CoDual{Symbol},
+    id::CoDual{Symbol},
+    file::CoDual{String},
+    line::CoDual{Int64},
+)
+    result = Core._call_latest(
+        Core.kwcall,
+        kwargs.x,
+        CoreLogging.handle_message,
+        logger.x,
+        loglevel.x,
+        message.x,
+        _module.x,
+        group.x,
+        id.x,
+        file.x,
+        line.x;
+    )
+    return zero_fcodual(result), NoPullback()
+end
+
 function generate_hand_written_rrule!!_test_cases(
     rng_ctor, ::Val{:avoiding_non_differentiable_code}
 )
     _x = Ref(5.0)
     _dx = Ref(4.0)
     test_cases = vcat(
         Any[
-        # Rules to avoid pointer type conversions.
+            # Rules to avoid pointer type conversions.
             (
-            true,
-            :stability_and_allocs,
-            nothing,
-            +,
-            CoDual(
-                bitcast(Ptr{Float64}, pointer_from_objref(_x)),
-                bitcast(Ptr{Float64}, pointer_from_objref(_dx)),
+                true,
+                :stability_and_allocs,
+                nothing,
+                +,
+                CoDual(
+                    bitcast(Ptr{Float64}, pointer_from_objref(_x)),
+                    bitcast(Ptr{Float64}, pointer_from_objref(_dx)),
+                ),
+                2,
             ),
-            2,
-        ),],
+
+            # Rules for handling Atomic read operations.
+            (
+                false,
+                :stability_and_allocs,
+                nothing,
+                Base.getindex,
```

Comment:
For consistency, let's use `Base.getindex` or `getindex` for all tests here. 
```suggestion
                getindex,
```